File: backend-nestjs/src/admin/admin.service.ts
import {
  ForbiddenException,
  Injectable,
  NotFoundException,
} from '@nestjs/common';
import { PaginationDto } from './dto/pagination.dto';
import { PrismaService } from 'src/prisma/prisma.service';

@Injectable()
export class AdminService {
  constructor(private prisma: PrismaService) {}
  async listUsers(dto: PaginationDto) {
    const page = dto.page ?? 1;
    const limit = dto.limit ?? 10;
    const skip = (page - 1) * limit;

    const usuarios = await this.prisma.user.findMany({
      skip: skip,
      take: limit,
      select: {
        id: true,
        name: true,
        email: true,
        role: true,
        twoFactorEnabled: true,
        createdAt: true,
        updatedAt: true,
      },
    });

    const total = await this.prisma.user.count();

    return {
      data: usuarios,
      total: total,
      page: page,
      limit: limit,
    };
  }

  async deleteUser(targetId: string, requesterId: string) {
    const targetUser = await this.prisma.user.findUnique({
      where: { id: targetId },
    });

    if (!targetUser) {
      throw new NotFoundException('Usuário não encontrado');
    }

    if (targetId === requesterId) {
      throw new ForbiddenException('Você não pode deletar sua própria conta');
    }

    if (targetUser.role === 'ADMIN') {
      throw new ForbiddenException(
        'Não é possível deletar outro administrador',
      );
    }

    await this.prisma.user.delete({ where: { id: targetId } });
    return { message: 'Usuário deletado com sucesso' };
  }
}

File: backend-springboot/backend-springboot/src/main/java/com/adryan/authbenchmark/backend_springboot/service/AdminService.java
package com.adryan.authbenchmark.backend_springboot.service;

import com.adryan.authbenchmark.backend_springboot.dto.PagedResponseDto;
import com.adryan.authbenchmark.backend_springboot.dto.UserResponseDto;
import com.adryan.authbenchmark.backend_springboot.exception.AutoExclusionException;
import com.adryan.authbenchmark.backend_springboot.exception.CannotDeleteAdminException;
import com.adryan.authbenchmark.backend_springboot.model.Role;
import com.adryan.authbenchmark.backend_springboot.model.User;
import com.adryan.authbenchmark.backend_springboot.repository.UserRepository;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.UUID;

@Service
public class AdminService {

    private final UserRepository userRepository;

    public AdminService(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    public PagedResponseDto<UserResponseDto> listUsers(int page, int limit) {
        Pageable pageable = PageRequest.of(page - 1, limit);
        Page<User> result = userRepository.findAll(pageable);

        List<UserResponseDto> data = result.getContent()
                .stream()
                .map(UserResponseDto::new)
                .toList();

        return new PagedResponseDto<>(
                data,
                result.getTotalElements(),
                result.getTotalPages(),
                page
        );
    }

    public void deleteUsers(UUID targetId, UUID requesterId){
        if(targetId.equals(requesterId)){
            throw new AutoExclusionException("Você não pode deletar sua própria conta");
        }

        User targetUser = userRepository.findById(targetId)
                .orElseThrow(() -> new RuntimeException("Usuário não encontrado"));

        if (targetUser.getRole() == Role.ADMIN) {
            throw new CannotDeleteAdminException("Não é possível deletar outro administrador");
        }

        userRepository.deleteById(targetId);
    }
}